core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1007,6 +1007,49 @@ public Builder setBranchSnapshot(long snapshotId, String branch) {
       return this;
     }
 
+    public Builder setTag(String tag, SnapshotRef ref) {
+      SnapshotRef existingRef = refs.get(tag);
+      if (existingRef != null && existingRef.equals(ref)) {
+        return this;
+      }
+      long snapshotId = ref.snapshotId();
+      Snapshot snapshot = snapshotsById.get(snapshotId);
+      ValidationException.check(snapshot != null, "Cannot set %s to unknown snapshot: %s", tag, snapshotId);
+      ValidationException.check(ref.isTag(), "Cannot create tag: %s is a branch", tag);
```

Comment:
What about `setRef` instead of checking that `setTag` was called with a tag? Is there a significant difference between how they are handled?

Reply:
There really isn't. We can handle it in the same method.